packages/git/src/browser/git-decorator.ts

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { inject, injectable } from 'inversify';
+import URI from '@theia/core/lib/common/uri';
+import { ILogger } from '@theia/core/lib/common/logger';
+import { Event, Emitter } from '@theia/core/lib/common/event';
+import { ITree, ITreeNode } from '@theia/core/lib/browser/tree/tree';
+import { TreeNodeIterator } from '@theia/core/lib/browser/tree/tree-iterator';
+import { DisposableCollection } from '@theia/core/lib/common/disposable';
+import { PreferenceChangeEvent } from '@theia/core/lib/browser/preferences/preference-proxy';
+import { TreeDecorator, TreeDecoration } from '@theia/core/lib/browser/tree/tree-decorator';
+import { Git } from '../common/git';
+import { GitWatcher } from '../common/git-watcher';
+import { WorkingDirectoryStatus } from '../common/git-model';
+import { GitRepositoryProvider } from './git-repository-provider';
+import { GitFileChange, GitFileStatus } from '../common/git-model';
+import { GitDecorationsPreferences, GitDecorationsConfiguration, GIT_DECORATIONS } from './git-decorator-preferences';
+
+@injectable()
+export class GitDecorator implements TreeDecorator {
+
+    readonly id = 'theia-git-decorator';
+
+    protected readonly toDispose: DisposableCollection;
+    protected readonly emitter: Emitter<(tree: ITree) => Map<string, TreeDecoration.Data>>;
+
+    protected enabled: boolean;
+    protected showColors: boolean;
+
+    constructor(
+        @inject(Git) protected readonly git: Git,
+        @inject(GitRepositoryProvider) protected readonly repositoryProvider: GitRepositoryProvider,
+        @inject(GitWatcher) protected readonly watcher: GitWatcher,
+        @inject(GitDecorationsPreferences) protected readonly preferences: GitDecorationsPreferences,
+        @inject(ILogger) protected readonly logger: ILogger) {
+        this.emitter = new Emitter();
+        this.toDispose = new DisposableCollection();
+        repositoryProvider.onDidChangeRepository(async repository => {
+            this.toDispose.dispose();
+            if (repository) {
+                this.toDispose.pushAll([
+                    await this.watcher.watchGitChanges(repository),
+                    this.watcher.onGitEvent(event => this.fireDidChangeDecorations((tree: ITree) => this.collectDecorators(tree, event.status)))
+                ]);
+            }
+        });
+        this.preferences.onPreferenceChanged(event => this.handlePreferenceChange(event));
+        this.enabled = this.preferences[GIT_DECORATIONS.ENABLED];
```

Comment:
why is it needed? ts will report an error if you type something else

an idea of having preference proxies to have statically typed apis, in this case the string is a function name